kinesumer.go

```diff
@@ -376,12 +410,24 @@ func (k *Kinesumer) deregisterConsumers() {
 func (k *Kinesumer) start() {
 	k.stop = make(chan struct{})
 
-	if k.efoMode {
-		k.consumeEFOMode()
-	} else {
-		k.consumePolling()
+	for stream, shardsPerStream := range k.shards {
+		for _, shard := range shardsPerStream {
+			k.wait.Add(1)
+			k.startConsuming(stream, shard)
+		}
 	}
+}
 
+func (k *Kinesumer) startConsuming(stream string, shard *Shard) {
+	if k.efoMode && k.autoCommitEnabled {
+		go k.consumePipeWithAutoCommit(stream, shard)
+	} else if k.efoMode && !k.autoCommitEnabled {
+		go k.consumePipeWithManualCommit(stream, shard)
+	} else if !k.efoMode && k.autoCommitEnabled {
+		go k.consumePollingWithAutoCommit(stream, shard)
+	} else if !k.efoMode && !k.autoCommitEnabled {
+		go k.consumePollingWithManualCommit(stream, shard)
+	}
```

Comment:
Personally, I don't like this approach, and I don't think it's a good idea to duplicate the entire consumer logic by how we commit the sequence numbers. 

Consuming and committing are done independently, so they don't need to be coupled as an entire code block. This means we can update the commit logics to handle both cases without re-write entire consume methods.

The simplest approach is that spawn a goroutine doing commit only if auto commit option is enabled, do nothing otherwise. I think it would be better than re-write entire logics.

Reply:
Thanks, that's brilliant insight!